### PR TITLE
Unfinished controller support, please follow comments

### PR DESCRIPTION
**Pull request by Arman-AIi**
note: this has only been tested on stadia controller
changes are only in client > src > scripts > managers > inputManager.ts

### CODE REVIEW

**Review by Compositr — COMMENTED** [review]
Is there any way to do this without having a ton of gamepadX or whatever vars?

**Comment by Arman-AIi** [comment]
You can condense them into a few lines but you need them to use all controller buttons

**Comment by tomatopotato17265** [comment]
Controls recommendations:

B: drop items (press to scroll between items, press and hold to actually drop)
X: map
Y: Special emotes (ammo request, medkit request, etc.)

idk how this works but if you can take advantage of the mini-touchpad on ps4 and ps5 controllers for a mini cursor some ppl might like that

**Comment by Arman-AIi** [comment]
> Controls recommendations:
> 
> B: drop items (press to scroll between items, press and hold to actually drop) X: map Y: Special emotes (ammo request, medkit request, etc.)
> 
> idk how this works but if you can take advantage of the mini-touchpad on ps4 and ps5 controllers for a mini cursor some ppl might like that

how will Y for special emotes work exactly

**Comment by tomatopotato17265** [comment]
> > Controls recommendations:
> 
> > 
> 
> > B: drop items (press to scroll between items, press and hold to actually drop) X: map Y: Special emotes (ammo request, medkit request, etc.)
> 
> > 
> 
> > idk how this works but if you can take advantage of the mini-touchpad on ps4 and ps5 controllers for a mini cursor some ppl might like that
> 
> 
> 
> how will Y for special emotes work exactly

You hold down Y while using the D-pad. This system works similar to mobile special emotes: you toggle that cellular tower thing then select the emotes.

**Comment by Arman-AIi** [comment]
> > > Controls recommendations:
> > 
> > 
> > > 
> > 
> > 
> > > B: drop items (press to scroll between items, press and hold to actually drop) X: map Y: Special emotes (ammo request, medkit request, etc.)
> > 
> > 
> > > 
> > 
> > 
> > > idk how this works but if you can take advantage of the mini-touchpad on ps4 and ps5 controllers for a mini cursor some ppl might like that
> > 
> > 
> > how will Y for special emotes work exactly
> 
> You hold down Y while using the D-pad. This system works similar to mobile special emotes: you toggle that cellular tower thing then select the emotes.

Sounds good. The primary goal is to have bindable controller buttons and working analog sticks, defaults are just a nice-to-have. We also must have a controller-navigable UI

**Comment by tomatopotato17265** [comment]
> > > > Controls recommendations:
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > B: drop items (press to scroll between items, press and hold to actually drop) X: map Y: Special emotes (ammo request, medkit request, etc.)
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > idk how this works but if you can take advantage of the mini-touchpad on ps4 and ps5 controllers for a mini cursor some ppl might like that
> 
> > > 
> 
> > > 
> 
> > > how will Y for special emotes work exactly
> 
> > 
> 
> > You hold down Y while using the D-pad. This system works similar to mobile special emotes: you toggle that cellular tower thing then select the emotes.
> 
> 
> 
> Sounds good. The primary goal is to have bindable controller buttons and working analog sticks, defaults are just a nice-to-have. We also must have a controller-navigable UI

We should take things one at a time. Suroi for console doesn't exist yet so we don't need to worry about controller-navigable UI. I say we first add the functionality with default controls, then add the ability to customized bindings. Controller-navigable UI is something we don't have to worry about until we make Suroi for Xbox, PlayStation, Nintendo Switch, or other controller-based platform.

**Comment by kenos1** [comment]
this easily has to be one of the most atrocious pieces of code i've ever seen from a suroi dev

how hard is it to not repeat yourself and do something like

```js
const GamepadButtonMap = ["GamepadA", ...] // truncated for demo purposes

if (event instanceof GamepadEvent) {
  input = GamepadButtonMap.find((_, i) => event.gamepad.buttons[i].pressed) ?? "GamepadUnknown"
}
```

File: client/src/scripts/managers/inputManager.ts [hunk 1]
```diff
@@ -16,6 +16,92 @@ import { FIRST_EMOTE_ANGLE, FOURTH_EMOTE_ANGLE, PIXI_SCALE, SECOND_EMOTE_ANGLE,
...
+        console.log(gamepadLeftStickVertical);
+        console.log(gamepadRightStickHorizontal);
+        console.log(gamepadRightStickVertical);
+        /*honestly you could just derive the controls from mobile and add 
+        the physical joysticks as input instead of the virtual joysticks */
+    }
+}
+
+
+function gameLoop(){
```
Comment (kenos1) [inline client/src/scripts/managers/inputManager.ts:96]: me when PIXI.Application.ticker

**Review by Compositr — CHANGES_REQUESTED** [review]
We need to make this far, far more concise. Also what the hell is that gameLoop function... (see Kenos' comment)

**Comment by kenos1** [comment]
Also please read https://developer.mozilla.org/en-US/docs/Web/API/Gamepad/buttons as you are missing critical information lol

File: client/src/scripts/managers/inputManager.ts [hunk 2]
```diff
@@ -16,6 +16,92 @@ import { FIRST_EMOTE_ANGLE, FOURTH_EMOTE_ANGLE, PIXI_SCALE, SECOND_EMOTE_ANGLE,
 import { getTranslatedString } from "../../translations";
```
Comment (kenos1) [inline client/src/scripts/managers/inputManager.ts:16]: oh yeah by the way to anyone whos reading patch-notes channel i am homeless and also the fbi is after me

**Review by kenos1 — APPROVED** [review]
oh yeah by the way to anyone whos reading patch-notes channel i am homeless and also the fbi is after me

**Comment by Arman-AIi** [comment]
> Also please read https://developer.mozilla.org/en-US/docs/Web/API/Gamepad/buttons as you are missing critical information lol

This is the guide I am using

**Review by leia-uwu — APPROVED** [review]
hiiii suroi patch-notes channel I'm totally original and not copying kenos
anyways LGTM should merge asap and push to production :+1:

**Comment by Arman-AIi** [comment]
> hiiii suroi patch-notes channel I'm totally original and not copying kenos anyways LGTM should merge asap and push to production 👍

I agree with this information

**Comment by Arman-AIi** [comment]
> We need to make this far, far more concise. Also what the hell is that gameLoop function... (see Kenos' comment)

The requested changes have been fully completed sir.


also how did I wake up leia and kenos 💀 💀

**Comment by leia-uwu** [comment]
> also how did I wake up leia and kenos 💀 💀

kenos pinged me to make fun of your code

File: client/src/scripts/managers/inputManager.ts [hunk 3]
```diff
@@ -485,7 +537,34 @@ export class InputManager {
 
             return input as Ret;
         }
+        if (event instanceof GamepadEvent) {
+            gameLoop();
```
Comment (ei-pi) [inline client/src/scripts/managers/inputManager.ts:541]: kill kill kill kill kill kill kill kill kill kill kill kill kill kill kill kill kill kill kill kill kill kill kill kill kill kill kill kill

**Comment by 1092384** [comment]
> You can condense them into a few lines but you need them to use all controller buttons

i've decided you're a furry

**Comment by kenos1** [comment]
image

**Comment by Arman-AIi** [comment]
I already changed it why so mean 😢

**Review by tomatopotato17265 — COMMENTED** [review]
While you're decluttering the code, pls add comments for noobs like me :)

**Comment by ersek-huba** [comment]
This is the funniest PR comment section, I've ever seen so far (after the Chinese John Cena one (including the Linux PR)).

**Comment by Arman-AIi** [comment]
new commit, fixed issue where the buttons just refused to register, working and tested

**Pull request closed without merging**